Communication/Data_transferring/TransferData.cpp

```diff
@@ -0,0 +1,148 @@
+        bool result = future.get();
+        std::cout << "Result: " << std::boolalpha << result << std::endl;
+    }
+}
+bool TransferData::sendMessageByChunk(const string& chunk)
+{
+    try
+    {
+        if (chunk.empty()) {
+            throw std::runtime_error("Chunk is empty");
+        }
+        //Noa function, open the socket.
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        std::cout << "Sent chunk: " << chunk << std::endl;
+        return true;
+    }
+    catch (const std::exception&)
+    {
+        throw exception("The send failed");
+    }
+
+}
+bool TransferData::sendMetaData(const Meta_Data& metaData)
+{
+    // ��� ����� �� ���� ������
+    try
+    {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100)); // ����� ��� �����
+        std::cout << "Sent metaData validation: " << std::endl; //<< metaData 
+        return true;
+    }
+    catch (const std::exception&)
+    {
+        throw exception("The send failed");
+    }
+
+}
+bool TransferData::sendData(const string& data, const Meta_Data& metaData) {
+    try
+    {
+        if (data.empty()) {
+            throw std::runtime_error("Data string is empty");
+        }
+        //Noa function
+        return true;
+    }
+    catch (const std::exception&)
+    {
+        throw exception("The send failed");
+    }
+}
+void TransferData::SendsAsynchronously(const string& str, const Meta_Data& metaData, size_t numChunks, size_t chunk_size, size_t numThreads)
+{
+    if (str.empty()) {
+        throw std::runtime_error("Data is empty");
+    }
+    size_t str_size = str.size();
+    //open the threads and send to socket function.
+    vector<future<bool>> futures;
+    size_t chunkStart = 0;
+    for (size_t i = 0; i < numChunks; ++i) {
+        size_t chunkEnd = min(chunkStart + chunk_size, str_size);
+        string chunk = str.substr(chunkStart, chunkEnd - chunkStart);
+        chunkStart = chunkEnd;
+        // If we have reached the maximum number of threads, wait for one to finish
+        if (futures.size() >= numThreads) {
+            for (auto it = futures.begin(); it != futures.end();) {
+                if (it->wait_for(chrono::seconds(0)) == future_status::ready) {
+                    it = futures.erase(it);
+                }
+                else {
+                    ++it;
+                }
+            }
+        }
+        // Submit the block in a new thread
+        futures.push_back(async(launch::async, &TransferData::sendMessageByChunk, this, chunk));
+    }
+    futures.push_back(async(launch::async, &TransferData::sendMetaData, this, std::cref(metaData)));
+
+    // Wait for all remaining threads to finish
+    waiting(futures);
+}
+void TransferData::SendsSynchronously(const string& str, const Meta_Data& metaData)
+{
+    if (str.empty()) {
+        throw std::runtime_error("Data is empty");
+    }
+    this->sendData(str, metaData);
+}
+int TransferData::choosing_an_option_to_transfer() {
```

Comment:
I think you should use a define in the TransferData.h file and put it on for now, 
in this function you need only to check the variable 
I dont think that in real life the operator will need to type his choosing

Reply:
You said to give a choice, didn't you?
How will it work?
Currently hide this function and set it to always be asynchronous?